File: frontend/lib/semantic/formula.ts
/**
 * Block-based formula model for computed metrics and KPI calculations.
 *
 * The dialog edits a FormulaGroup (a flat row of terms joined by operators,
 * with nested groups for parentheses). It serializes to the recursive AST
 * stored on the backend (see backend app/modules/semantic/formula.py) with
 * standard precedence: × and ÷ bind before + and −.
 */

// 'none' = the raw column value (no aggregation). A formula that uses any
// 'none' term is row-level and behaves like a raw measure.
export type FormulaAggName = 'sum' | 'avg' | 'min' | 'max' | 'count' | 'count_distinct' | 'none'
export type FormulaOp = '+' | '-' | '*' | '/'

export type FormulaNode =
  | { type: 'binary'; op: FormulaOp; left: FormulaNode; right: FormulaNode }
  | { type: 'agg'; agg: FormulaAggName; column: string }
  | { type: 'literal'; value: number }
  | { type: 'ref'; measure: string } // KPI mode only

export type FormulaTerm =
  | { kind: 'agg'; agg: FormulaAggName; column: string }
  | { kind: 'ref'; measure: string } // KPI mode only
  | { kind: 'literal'; value: number }
  | { kind: 'group'; group: FormulaGroup } // explicit parentheses

export interface FormulaGroup {
  terms: FormulaTerm[]
  ops: FormulaOp[] // ops.length === terms.length - 1
}

export const FORMULA_AGGS: FormulaAggName[] = ['sum', 'avg', 'min', 'max', 'count', 'count_distinct', 'none']
// Dropdown labels. "none" is the origin/raw value (no aggregation).
export const AGG_LABEL: Record<FormulaAggName, string> = {
  sum: 'SUM',
  avg: 'AVG',
  min: 'MIN',
  max: 'MAX',
  count: 'COUNT',
  count_distinct: 'COUNT DISTINCT',
  none: 'Raw value (no agg)',
}
export const OP_SYMBOL: Record<FormulaOp, string> = { '+': '+', '-': '−', '*': '×', '/': '÷' }

function termToAst(term: FormulaTerm): FormulaNode {
  if (term.kind === 'agg') return { type: 'agg', agg: term.agg, column: term.column }
  if (term.kind === 'ref') return { type: 'ref', measure: term.measure }
  if (term.kind === 'literal') return { type: 'literal', value: term.value }
  return groupToAst(term.group)
}

/** Serialize an edited group to the AST, folding × ÷ before + −. */
export function groupToAst(group: FormulaGroup): FormulaNode {
  const terms = group.terms.map(termToAst)
  if (!terms.length) throw new Error('Empty formula')

  // First pass: fold multiplicative runs left-to-right.
  const nodes: FormulaNode[] = [terms[0]]
  const addOps: FormulaOp[] = []
  for (let i = 0; i < group.ops.length; i++) {
    const op = group.ops[i]
    const term = terms[i + 1]
    if (op === '*' || op === '/') {
      nodes[nodes.length - 1] = { type: 'binary', op, left: nodes[nodes.length - 1], right: term }
    } else {
      nodes.push(term)
      addOps.push(op)
    }
  }
  // Second pass: fold additive runs left-to-right.
  let ast = nodes[0]
  for (let i = 0; i < addOps.length; i++) {
    ast = { type: 'binary', op: addOps[i], left: ast, right: nodes[i + 1] }
  }
  return ast
}

function leafOrGroup(node: FormulaNode): FormulaTerm {
  // A binary node in a leaf slot is always an explicit parenthesized group —
  // operator chains at the current level were already flattened by the caller.
  if (node.type === 'binary') return { kind: 'group', group: astToGroup(node) }
  if (node.type === 'agg') return { kind: 'agg', agg: node.agg, column: node.column }
  if (node.type === 'ref') return { kind: 'ref', measure: node.measure }
  return { kind: 'literal', value: node.value }
}

/** Inverse of groupToAst — flatten an AST back into editable rows. */
export function astToGroup(node: FormulaNode): FormulaGroup {
  const terms: FormulaTerm[] = []
  const ops: FormulaOp[] = []

  function pushMul(n: FormulaNode) {
    if (n.type === 'binary' && (n.op === '*' || n.op === '/')) {
      pushMul(n.left)
      ops.push(n.op)
      terms.push(leafOrGroup(n.right))
    } else {
      terms.push(leafOrGroup(n))
    }
  }

  function pushAdd(n: FormulaNode) {
    if (n.type === 'binary' && (n.op === '+' || n.op === '-')) {
      pushAdd(n.left)
      ops.push(n.op)
      pushMul(n.right)
    } else {
      pushMul(n)
    }
  }

  pushAdd(node)
  return { terms, ops }
}

/** Convert a legacy ratio expression for editing in the block builder. */
export function ratioToGroup(expression: Record<string, unknown>): FormulaGroup {
  const num = (expression.numerator ?? {}) as { agg?: string; column?: string }
  const den = (expression.denominator ?? {}) as { agg?: string; column?: string }
  const group: FormulaGroup = {
    terms: [
      { kind: 'agg', agg: (num.agg as FormulaAggName) || 'sum', column: num.column || '' },
      { kind: 'agg', agg: (den.agg as FormulaAggName) || 'sum', column: den.column || '' },
    ],
    ops: ['/'],
  }
  const multiplier = Number(expression.multiplier ?? 1)
  if (multiplier && multiplier !== 1) {
    group.terms.push({ kind: 'literal', value: multiplier })
    group.ops.push('*')
  }
  return group
}

/**
 * Evaluate a ref/literal/binary formula over already-aggregated values
 * (KPI custom calculations). Division by zero and missing refs yield null.
 */
export function evaluateFormula(
  node: FormulaNode,
  refs: Record<string, number | null | undefined>,
): number | null {
  if (node.type === 'binary') {
    const left = evaluateFormula(node.left, refs)
    const right = evaluateFormula(node.right, refs)
    if (left == null || right == null) return null
    if (node.op === '+') return left + right
    if (node.op === '-') return left - right
    if (node.op === '*') return left * right
    return right === 0 ? null : left / right
  }
  if (node.type === 'literal') return node.value
  if (node.type === 'ref') {
    const value = refs[node.measure]
    return value == null ? null : Number(value)
  }
  return null // agg nodes are not evaluable over query results
}

/** Human-readable rendering of a group, e.g. (SUM(Revenue) − SUM(Spend)) ÷ SUM(Spend) × 100 */
export function formatGroup(group: FormulaGroup, labelFor: (name: string) => string): string {
  const parts: string[] = []
  group.terms.forEach((term, i) => {
    if (i > 0) parts.push(OP_SYMBOL[group.ops[i - 1]])
    if (term.kind === 'agg') {
      parts.push(term.agg === 'none' ? labelFor(term.column) : `${term.agg.toUpperCase()}(${labelFor(term.column)})`)
    } else if (term.kind === 'ref') parts.push(labelFor(term.measure))
    else if (term.kind === 'literal') parts.push(String(term.value))
    else parts.push(`( ${formatGroup(term.group, labelFor)} )`)
  })
  return parts.join(' ')
}

/** True when every term is filled in and the group is non-empty. */
export function isGroupComplete(group: FormulaGroup): boolean {
  if (!group.terms.length) return false
  return group.terms.every((term) => {
    if (term.kind === 'agg') return !!term.column
    if (term.kind === 'ref') return !!term.measure
    if (term.kind === 'literal') return Number.isFinite(term.value)
    return isGroupComplete(term.group)
  })
}

File: backend/app/modules/semantic/formula.py
"""
app/modules/semantic/formula.py
=================================
Custom computed-metric formulas: a small recursive expression AST stored as
JSON (structured — never raw SQL from the user) and compiled to SQL by both
engines (Cube YAML and the local DuckDB backend).

Stored shape:  {"type": "formula", "root": <node>}

  node := {"type": "binary", "op": "+"|"-"|"*"|"/", "left": node, "right": node}
        | {"type": "agg", "agg": "sum"|"avg"|"min"|"max"|"count"|"count_distinct"|"none",
           "column": "<semantic column>"}
        | {"type": "literal", "value": <number>}
        | {"type": "ref", "measure": "<member name>"}   # KPI definitions only

"none" means the raw column value (no aggregation). A formula that uses any
"none" term is *row-level*: it produces one value per row, runs only on the
DuckDB detail path, and cannot be mixed with real aggregations (SUM/AVG/…) —
just like a raw column measure. Everything else is a normal aggregate metric.

The engines differ only in how a column name resolves to a quoted SQL
identifier, so the compiler takes a resolve_column callback. Division always
compiles to NULLIF-guarded SQL; parenthesization is inherent to the tree.
"""
from __future__ import annotations

from typing import Annotated, Callable, Iterator, Literal, Mapping, Union

from pydantic import BaseModel, Field, TypeAdapter, ValidationError

FORMULA_AGGS = ("sum", "avg", "min", "max", "count", "count_distinct", "none")
# A flat N-term row folds into a left-deep binary tree of depth ~N, so the
# depth cap must comfortably exceed the per-row term count; MAX_NODES is the
# real size limit, depth only guards recursion.
MAX_DEPTH = 32
MAX_NODES = 50


class FormulaError(ValueError):
    """A formula failed validation or compilation."""


class FormulaAgg(BaseModel):
    type: Literal["agg"]
    agg: Literal["sum", "avg", "min", "max", "count", "count_distinct", "none"]
    column: str


class FormulaLiteral(BaseModel):
    type: Literal["literal"]
    value: float


class FormulaRef(BaseModel):
    type: Literal["ref"]
    measure: str


class FormulaBinary(BaseModel):
    type: Literal["binary"]
    op: Literal["+", "-", "*", "/"]
    left: "FormulaNode"
    right: "FormulaNode"


FormulaNode = Annotated[
    Union[FormulaBinary, FormulaAgg, FormulaLiteral, FormulaRef],
    Field(discriminator="type"),
]

FormulaBinary.model_rebuild()

_NODE_ADAPTER: TypeAdapter = TypeAdapter(FormulaNode)

_Node = Union[FormulaBinary, FormulaAgg, FormulaLiteral, FormulaRef]


def parse_formula(root: dict) -> _Node:
    """Parse + structurally validate a formula node dict. Raises FormulaError."""
    if not isinstance(root, dict):
        raise FormulaError("Formula root must be an object.")
    try:
        return _NODE_ADAPTER.validate_python(root)
    except ValidationError as e:
        first = e.errors()[0] if e.errors() else {}
        raise FormulaError(f"Malformed formula: {first.get('msg', 'invalid node')}") from e


def _walk(node: _Node, depth: int = 1) -> Iterator[tuple[_Node, int]]:
    yield node, depth
    if isinstance(node, FormulaBinary):
        yield from _walk(node.left, depth + 1)
        yield from _walk(node.right, depth + 1)


def iter_formula_columns(root: dict) -> Iterator[str]:
    """Yield every column referenced by agg nodes in a formula."""
    for node, _ in _walk(parse_formula(root)):
        if isinstance(node, FormulaAgg):
            yield node.column


def formula_is_row_level(root: dict) -> bool:
    """True if any term uses the raw ("none") aggregation — a row-level formula."""
    for node, _ in _walk(parse_formula(root)):
        if isinstance(node, FormulaAgg) and node.agg == "none":
            return True
    return False


def expression_is_row_level(expression: dict | None) -> bool:
    """True if a stored metric expression is a row-level formula (raw values).

    Row-level metrics run only on the DuckDB detail path and behave like raw
    column measures; the Cube schema never receives them.
    """
    if not expression or expression.get("type") != "formula":
        return False
    try:
        return formula_is_row_level(expression.get("root") or {})
    except FormulaError:
        return False


def validate_formula(
    root: dict,
    valid_columns: set[str] | None = None,
    *,
    allow_refs: bool = False,
    valid_refs: set[str] | None = None,
) -> None:
    """
    Full validation of a formula: structure, size limits, column existence,
    and ref policy (refs are only legal in KPI definitions, where they must
    point at the query's selected measures). Raises FormulaError.
    """
    parsed = parse_formula(root)

    count = 0
    has_operand = False
    has_raw = False   # a "none" term (raw column value)
    has_agg = False   # a real aggregation (SUM/AVG/…)
    for node, depth in _walk(parsed):
        count += 1
        if count > MAX_NODES:
            raise FormulaError(f"Formula too large (max {MAX_NODES} terms).")
        if depth > MAX_DEPTH:
            raise FormulaError(f"Formula too deeply nested (max depth {MAX_DEPTH}).")
        if isinstance(node, FormulaAgg):
            has_operand = True
            if node.agg == "none":
                has_raw = True
            else:
                has_agg = True
            if valid_columns is not None and node.column not in valid_columns:
                raise FormulaError(f"Unknown column {node.column!r} in formula.")
        elif isinstance(node, FormulaRef):
            has_operand = True
            if not allow_refs:
                raise FormulaError("Measure references are not allowed in computed metrics.")
            if valid_refs is not None and node.measure not in valid_refs:
                raise FormulaError(f"Unknown measure {node.measure!r} in formula.")

    if not has_operand:
        raise FormulaError("Formula must reference at least one column or measure.")
    if has_raw and has_agg:
        raise FormulaError(
            "A formula can't mix raw column values with aggregations like SUM or AVG."
        )


def compile_formula_sql(
    root: dict,
    resolve_column: Callable[[str], str],
    *,
    allow_refs: bool = False,
    resolve_ref: Callable[[str], str] | None = None,
) -> str:
    """
    Compile a formula node to SQL. resolve_column maps a semantic column name
    to a quoted identifier (each engine supplies its own); resolve_ref maps a
    measure reference the same way when allow_refs is set.
    """
    def emit(node: _Node) -> str:
        if isinstance(node, FormulaBinary):
            left = emit(node.left)
            right = emit(node.right)
            if node.op == "/":
                return f"({left} / NULLIF({right}, 0))"
            return f"({left} {node.op} {right})"
        if isinstance(node, FormulaAgg):
            ident = resolve_column(node.column)
            if node.agg == "none":
                return ident  # raw column value (row-level term)
            if node.agg == "count_distinct":
                return f"COUNT(DISTINCT {ident})"
            return f"{node.agg.upper()}({ident})"
        if isinstance(node, FormulaLiteral):
            return repr(float(node.value))
        if not allow_refs or resolve_ref is None:
            raise FormulaError("Measure references are not allowed in computed metrics.")
        return resolve_ref(node.measure)

    return emit(parse_formula(root))


def evaluate_formula(root: dict, refs: Mapping[str, float | None]) -> float | None:
    """
    Evaluate a ref/literal/binary formula over already-aggregated values
    (KPI custom calculations). agg nodes cannot be evaluated here. Division
    by zero and missing refs yield None, mirroring SQL NULL semantics.
    """
    def ev(node: _Node) -> float | None:
        if isinstance(node, FormulaBinary):
            left = ev(node.left)
            right = ev(node.right)
            if left is None or right is None:
                return None
            if node.op == "+":
                return left + right
            if node.op == "-":
                return left - right
            if node.op == "*":
                return left * right
            return None if right == 0 else left / right
        if isinstance(node, FormulaLiteral):
            return float(node.value)
        if isinstance(node, FormulaRef):
            value = refs.get(node.measure)
            return None if value is None else float(value)
        raise FormulaError("Aggregations cannot be evaluated over query results.")

    return ev(parse_formula(root))
